File: questions/278.FirstBadVersion.js
/**
 * Definition for isBadVersion()
 *
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
const solution = function (isBadVersion) {
  /**
   * @param {integer} n Total versions
   * @return {integer} The first bad version
   */
  return function (n) {
    let start = 1
    let end = n
    let result = n

    while (end >= start) {
      const middle = Math.floor((start + end) / 2)
      if (isBadVersion(middle)) {
        result = middle
        end = middle - 1
      } else {
        start = middle + 1
      }
    }
    return result
  }
}

console.log(solution((n) => n === 4)(5))

File: questions/278.FirstBadVersion.go
package questions

import "math"

func firstBadVersion(n int) int {
	start := 1
	end := n
	result := n

	for end >= start {
		middle := int(math.Floor(float64(start+end) / 2))
		if isBadVersion(middle) {
			result = middle
			end = middle - 1
		} else {
			start = middle + 1
		}
	}
	return result
}

func isBadVersion(version int) bool {
	return version == 4
}
